### PR TITLE
Feat/new scg visualisation

### PR DESCRIPTION
**Pull request by ChaikaS**
This pull request is responsible for displaying elements in the SCg depending on their distance from the root element. Structure have to  keynodes with system identifiers rrel_main_key_sc_element or rrel_key_sc_element. Otherwise the displaying scg will occur normally.
For test:
- structure_IMS_user_interface
- section_type_purpose_and_structure_of_IMS
- section_objects_and_features_of_the_project_OSTIS

### CODE REVIEW

File: server/app.py [hunk 1]
```diff
@@ -23,168 +20,150 @@
...
-
-    tornado.options.define("apiai_subscription_key", default="", help="subscription key for api.ai", type=str)
-    tornado.options.define("apiai_client_access_token", default="", help="client access token for api.ai", type=str)
-
-    tornado.options.define("user_key_expire_time", default=600, help="user key expire time in seconds", type=int)
+    tornado.options.define(
+        "server_host", default="localhost", help="host name", type=str)
+    tornado.options.define("server_port", default=8090,
+                           help="host port", type=int)
+    # tornado.options.define(
```
Comment (smousy) [inline server/app.py:65]: why is it commented?

File: server/app.py [hunk 2]
```diff
@@ -23,168 +20,150 @@
...
         config.read(tornado.options.options.cfg)
-
     # prepare database
     database = db.DataBase()
     database.init()
-
     options_dict = tornado.options.options
-    server_url = f"ws://{options_dict.server_host}:{options_dict.server_port}/ws_json"
-
+    # server_url = f"ws://{options_dict.server_host}:{options_dict.server_port}/ws_json"
```
Comment (smousy) [inline server/app.py:96]: why is it commented?

File: server/app.py [hunk 3]
```diff
@@ -23,168 +20,150 @@
...
     # load scs required for sc-web server
     try:
         client.create_elements_by_scs(read_scs_fragments(REPO_FILE_PATH))
     except ServerError as e:
         logger.error(e)
         exit(1)
-
     ScKeynodes().resolve_identifiers([KeynodeSysIdentifiers])
-
+    # try:
```
Comment (smousy) [inline server/app.py:111]: why is it commented?

File: server/app.py [hunk 4]
```diff
@@ -217,32 +192,24 @@ def read_scs_fragments(root_path: str):
...
-
         else:
             continue
-
     return scs_fragments
-
-
 if __name__ == "__main__":
-    main()
+    main()
```
Comment (smousy) [inline server/app.py:215]: please add newline

**Review by smousy — CHANGES_REQUESTED** [review]

File: components/scg/src/scg-model-objects.js [hunk 5]
```diff
@@ -33,6 +33,23 @@ SCg.ModelObject = function (options) {
...
+    if (options.opacityElem) {
+        this.opacityElem = options.opacityElem;
+    } else {
+        this.opacityElem = 1;
+    }
+
+    if (options.widthEdge) {
+        this.widthEdge = options.widthEdge;
+    } else {
+        this.widthEdge = 8;
```
Comment (smousy) [inline components/scg/src/scg-model-objects.js:51]: why 8?

File: components/scg/src/scg-struct.js
Comment (smousy) [inline components/scg/src/scg-struct.js:1]: please change 1, 2, 3, 4 ... to something meaningful

File: components/scg/src/scg-model-objects.js [hunk 6]
```diff
@@ -60,6 +77,10 @@ SCg.ModelObject = function (options) {
     this.scene = null;
     this.bus = null;
     this.contour = null;
+    this.scaleElem = 1;
```
Comment (smousy) [inline components/scg/src/scg-model-objects.js:80]: please change number constants to something meaningful

File: client/js/Core/componentsandbox.js [hunk 7]
```diff
@@ -261,11 +263,261 @@ SCWeb.core.ComponentSandbox.prototype.createViewersForScStructs = function (cont
...
+        };
+        if (scene && self.isResultWithMainKey) {
+            let isEdge = scAddr ? scene.getObjectByScAddr(scAddr).edges[0].target : scene.getObjectByScAddr(resultLevelWithMainKey.get('mainNode').value).edges[0].target;
+            isEdge instanceof SCg.ModelEdge ? edgeToEdge = true : edgeToEdge = false;
+        };
+
+        let searchAllLevelEdges = async function (elementsArr, levelScales, level, visitedElements) {
+            let levelScale;
+
+            level > 6 ? levelScale = { node: 1, link: 1, opacity: 0.4, widthEdge: 6.5 } : levelScale = levelScales[level];
```
Comment (smousy) [inline client/js/Core/componentsandbox.js:349]: same

**Review by smousy — CHANGES_REQUESTED** [review]

**Review by smousy — COMMENTED** [review]
image

File: client/js/Core/componentsandbox.js [hunk 8]
```diff
@@ -261,16 +263,261 @@ SCWeb.core.ComponentSandbox.prototype.createViewersForScStructs = function (cont
...
+                        visitedElements.push(relationNode);
+                    }
+                    self.eventStructUpdate(true, triple.get("src").value, triple.get("edgeFromContourToRelationNode").value, scale);
+                    self.eventStructUpdate(true, triple.get("src").value, triple.get("edgeFromContourToEdgeFromRelationNodeToedgeFromMainNodeToSecondElem").value, scale);
+                };
+            };
+            return [...new Set(levelNodes)];
+        };
+
+        searchAllLevelEdges([mainElements], levelScales, 1, [...mainElements]);
```
Comment (smousy) [inline client/js/Core/componentsandbox.js:518]: why 1? whats that mean?

File: components/scg/src/scg-model-objects.js [hunk 9]
```diff
@@ -33,6 +33,23 @@ SCg.ModelObject = function (options) {
...
+    if (options.scaleElem) {
+        this.scaleElem = options.scaleElem;
+    } else {
+        this.scaleElem = 1;
+    }
+    
+    if (options.opacityElem) {
+        this.opacityElem = options.opacityElem;
+    } else {
+        this.opacityElem = 1;
```
Comment (smousy) [inline components/scg/src/scg-model-objects.js:45]: why 1?

File: components/scg/src/scg-struct.js [hunk 10]
```diff
@@ -54,14 +54,43 @@ function ScgFromScImpl(_sandbox, _editor, aMapping) {
...
 
-                if (editor.scene.getObjectByScAddr(addr))
-                    continue;
+                if (!task[2] && sandbox.mainElement) task[2] = { node: 1.8, link: 1.5, opacity: 1, widthEdge: 7.5, stroke: '#1E90FF', fill: '#1E90FF' };
+                if (!task[4] && sandbox.mainElement) task[4] = { node: 1.8, link: 1.5, opacity: 1, widthEdge: 7.5, stroke: '#1E90FF', fill: '#1E90FF' };
+
+                let newMainNode = editor.scene.getObjectByScAddr(addr);
+                if (newMainNode) {
+                    if (newMainNode instanceof SCg.ModelEdge) {
+                        newMainNode.setOpacityElem(task[4].opacity);
```
Comment (smousy) [inline components/scg/src/scg-struct.js:63]: why task 2,3,4? please add meaningful name

**Review by smousy — CHANGES_REQUESTED** [review]

**Review by smousy — CHANGES_REQUESTED** [review]
also you didn't fix previous comments

File: components/scg/src/scg-model-objects.js (hunk 6 above)
Comment (ChaikaS) [reply components/scg/src/scg-model-objects.js:80]: this is default value

File: client/js/Core/componentsandbox.js (hunk 8 above)
Comment (ChaikaS) [reply client/js/Core/componentsandbox.js:518]: find this function and see the arguments

File: components/scg/src/scg-model-objects.js (hunk 9 above)
Comment (ChaikaS) [reply components/scg/src/scg-model-objects.js:45]: this is default value

File: components/scg/src/scg-model-objects.js (hunk 5 above)
Comment (ChaikaS) [reply components/scg/src/scg-model-objects.js:51]: this is default value

**Review by smousy — APPROVED** [review]

**Comment by MikhailSadovsky** [comment]
Please update changelog with description what logic was added in this pr

**Review by MikhailSadovsky — APPROVED** [review]

**Pull request merged**